brave-intl/publishers
# Allow publishers to send themselves a new login link

### Changes
* Create custom failure app according to [this](https://github.com/plataformatec/devise/wiki/How-To:-Do-not-redirect-to-login-page-after-session-timeout) and [this](https://github.com/plataformatec/devise/wiki/How-To:-Redirect-to-a-specific-page-when-the-user-can-not-be-authenticated)
* Allow verified publishers to send themselves a new login link if they use an expired one
* Redirect session timeouts to login page
* Create new view for "send me login link page"

![image](https://user-images.githubusercontent.com/12549658/34583620-6d57a50e-f166-11e7-963c-c1d04dd7ecfd.png)

Resolves #381 

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave-intl/publishers/issues) for my issue if one did not already exist.
- [x] Tagged reviewers.

Test Plan:

I modified exisiting tests that assumed attempts to claim expired auth tokens would take them to the dashboard.

I created three new tests that verify that:
1. unverified publishers attempting to use expired token are still redirected to the dashboard with a flash message
2. verified publishers are are shown the "send me a new login link" page when they attempt to use an expired token
3. publishers whose session has expired will be sent to the login page with a flash message

1 & 2 are easy to verify through the browser:
1. create a new publisher, expire their token before clicking the verification link, then click the link.  This should bring you to the homepage.
2. create a new publisher, and click the link in the verification email.  Now click that link again, you should be brought to the "send me a new login link" page and send a valid link from there.

Reviewer Checklist:

Tests

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))
